packages/config-loader/src/schema/compile.ts

```diff
@@ -101,17 +158,68 @@ export function compileConfigSchemas(
 
   const merged = mergeConfigSchemas(schemas.map(_ => _.value));
 
-  if (options?.noUndeclaredProperties) {
-    traverse(merged, (schema: SchemaObject) => {
-      /**
-       * The `additionalProperties` key can only be applied to `type: object` in the JSON
-       *  schema.
-       */
-      if (schema?.type === 'object') {
-        schema.additionalProperties ||= false;
+  traverse(
+    merged,
+    (
+      schema: SchemaObject,
+      jsonPtr: string,
+      rootSchema: any,
+      _: any,
+      _1?: any,
+      parentSchema?: SchemaObject,
+    ) => {
+      // Inherit parent deepVisibility if we don't define one ourselves.
+      if (parentSchema?.deepVisibility) {
+        schema.deepVisibility ??= parentSchema?.deepVisibility;
+
+        /**
+         * Validate that we're not trying to override a child's visibility
+         *  by setting the parent deep visibility.
+         */
+        const values = [schema.visibility, parentSchema.visibility];
+        const hasFrontend = values.some(e => e === 'frontend');
+        const hasSecret = values.some(e => e === 'secret');
+        if (hasFrontend && hasSecret) {
+          throw new Error(
+            `Config schema visibility is both 'frontend' and 'secret' for ${jsonPtr}`,
+          );
+        }
+
+        // This runs before we compile the AJV so visibilityByDataPath has the
+        //  correct data.
+        schema.visibility ??= schema.deepVisibility;
+
+        /**
+         * Validate that we're not getting around secret visibility or frontend visibility by setting
+         *  some path frontend -> backend -> secret where we can't check for a simple switch from frontend
+         *  to secret visibility.
+         */
+        const visibilitiesFromRootToSchema = collectUniqueValuesForPath(
```

Comment:
Does make me wonder if there is a neater way to do this. Doesn't matter too much at this point though tbh, check is there and I bet it'll work well in practice. If possible we can tweak it in the future :+1:

Reply:
Probably is 😃 , should be easy to tweak in future.